File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& arr) {
        int len = arr.size();
        if(len <= 2) return 0;

        vector<int> left(len, 0), right(len, 0);
        left[0] = arr[0];
        right[len-1] = arr[len-1];

        for(int i = 1; i < len; i++){
            left[i] = max(left[i-1], arr[i]);
            right[len-i-1] = max(right[len-i], arr[len-i-1]);
        }

        int count = 0;
        for(int i = 0; i < len; i++){
            count += min(left[i], right[i]) - arr[i];
        }
        return count;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    // Time: O(n), Space: O(1)
    public static int trap(int[] height) {
        int len = height.length;
        int left = 0, right = len-1;
        int res = 0;
        int maxLeft = height[left], maxRight = height[right];
        while (left < right) {
            if (maxLeft < maxRight) {
                left++;
                maxLeft = Math.max(maxLeft, height[left]);
                res += maxLeft - height[left];
            } else {
                right--;
                maxRight = Math.max(maxRight, height[right]);
                res += maxRight - height[right];
            }
        }
        return res;
    }
}